infrap4d/infrap4d_main.cc

```diff
@@ -2,48 +2,51 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #if defined(DPDK_TARGET)
-#include "stratum/hal/bin/tdi/dpdk/dpdk_main.h"
-#elif defined (TOFINO_TARGET)
-#include "stratum/hal/bin/tdi/tofino/tofino_main.h"
+  #include "stratum/hal/bin/tdi/dpdk/dpdk_main.h"
+#else
+  #error "TDI target type not defined!"
 #endif
 
 extern "C"  {
 #include "daemon/daemon.h"
 }
 
-#include "krnlmon_main.h"
+#include "absl/synchronization/notification.h"
 #include "gflags/gflags.h"
+#include "krnlmon_main.h"
+#include "stratum/glue/status/status.h"
 
 DEFINE_bool(detach, true, "Run infrap4d in attached mode");
 
-pthread_cond_t rpc_start_cond = PTHREAD_COND_INITIALIZER;
-pthread_mutex_t rpc_start_lock = PTHREAD_MUTEX_INITIALIZER;
-int rpc_start_cookie = 0;
-
-pthread_cond_t rpc_stop_cond = PTHREAD_COND_INITIALIZER;
-pthread_mutex_t rpc_stop_lock = PTHREAD_MUTEX_INITIALIZER;
-int rpc_stop_cookie = 0;
+static inline ::util::Status target_main(
+    int argc, char** argv, absl::Notification * ready_sync,
+    absl::Notification* done_sync) {
+#if defined(DPDK_TARGET)
+  return stratum::hal::tdi::DpdkMain(argc, argv, ready_sync, done_sync);
+#else
+  #error "TDI target type not defined!"
+#endif
+}
 
 int main(int argc, char* argv[]) {
-#if defined(TOFINO_TARGET)
-  return stratum::hal::tdi::TofinoMain(argc, argv).error_code();
-#elif defined(DPDK_TARGET)
   gflags::ParseCommandLineFlags(&argc, &argv, false);
+
   if (FLAGS_detach) {
       daemonize_start(false);
       daemonize_complete();
   }
 
-  krnlmon_init();
+  absl::Notification ready_sync;
+  absl::Notification done_sync;
 
-  krnlmon_shutdown();
+  krnlmon_create_main_thread(&ready_sync);
+  krnlmon_create_shutdown_thread(&done_sync);
```

Comment:
Why do we need this shutdown to be called immediately after creating a switchlink thread?

Reply:
This function doesn't shut down; it creates a thread that _waits_ for a signal indicating that DpdkHal has shut down. That's why I renamed the function called here from "shutdown" to "create shutdown thread."

Unless I'm greatly mistaken, the existing shutdown thread has a bug in it. I'm pretty sure we should be passing `main_tid` to `switchlink_stop` so the latter can use the pid to stop the main switchlink thread. (Instead, it's using an uninitialized variable as the thread ID.)

I elected not to fix this in the initial Draft PR in order to focus on using dependency inversion to break the tight coupling between DpdkHal and Krnlmon.